src/types/Environment.cpp

```diff
@@ -9,30 +9,42 @@ using namespace ::Microsoft::Console::Utils;
 // We cannot use spand or not_null because we're dealing with \0\0-terminated buffers of unknown length
 #pragma warning(disable : 26481 26429)
 
+// Function Description:
+// - Wraps win32's CreateEnvironmentBlock to return a smart pointer.
+EnvironmentBlockPtr Microsoft::Console::Utils::CreateEnvironmentBlock()
+{
+    void* newEnvironmentBlock{ nullptr };
+    if (!::CreateEnvironmentBlock(&newEnvironmentBlock, GetCurrentProcessToken(), FALSE))

```

Comment:
@nathpete can you test what happens if you replace `GetCurrentProcessToken` with..


```c++
auto processToken{ wil::open_current_access_token() };
CreateEnv(...processToken.get()...)
```

If that doesn't restore `USERNAME`, can you try passing `TOKEN_QUERY|TOKEN_DUPLICATE` as the first arg to open_current_acc?

Reply:
whoops - @nathpete-msft

Reply:
```
auto processToken{ wil::open_current_access_token(TOKEN_QUERY | TOKEN_DUPLICATE) };
```
That does it!

Reply:
Ugh, I am surprised CreateEnvironmentBlock behaves that way. Tested:

- `NULL` => USERNAME=SYSTEM
- `GetCurrentProcessToken()` => no USERNAME
- `GetCurrentThreadEffectiveToken()` => no USERNAME
- `(HANDLE)0x123456` => error 6

Perhaps it is because GetCurrentProcessToken() lacks TOKEN_DUPLICATE access.

Reply:
You know, you're probably right about that last bit. That makes sense.

At least there's a nice WIL helper for us to get our current effective access token :smile: